deepspeed/runtime/engine.py

```diff
@@ -1208,13 +1208,9 @@ def _do_optimizer_sanity_check(self, basic_optimizer):
                         "**** You are using ZeRO with an untested optimizer, proceed with caution *****"
                     )
 
-            if model_dtype == torch.bfloat16 and grad_accum_dtype == torch.float32 and self.zero_optimization_stage(
-            ) == 1:
-                return BFLOAT16
-
-            if model_dtype != grad_accum_dtype:
+            if model_dtype != grad_accum_dtype and self.zero_optimization_stage() == 3:
```

Comment:
This change makes bf16_optimizer unusable, right?

Reply:
Correct, I believe it replaces all use cases of the current BF16 Optimizer.

Reply:
@stas00, what do you think? We probably need more testing to validate correctness parity before disabling bf16_optimizer. 

Also, bf16_optimizer has superior memory efficiency, and probably throughput. We need to validate that as well.

Reply:
I'm not sure what you are asking, Tunji. 

We know for sure that some people out there use Megatron-Deepspeeed (as we get Issues), so it's almost certain someone will want the BF16Optimizer in a working condition. If that's the question.

Reply:
Yes, that is the question. How to ensure BF16Optimizer users retain access?

Reply:
Sorry for not being clear. Let me back up a bit. With this PR, both zero_stage_1 now supports `model_dtype=bf16` and `accum_dtype=fp32` similar to bf16_optimizer.  And so, with that dtypes combination and `stage=1`, zero_stage_1 optimizer will now be instantiated instead of bf16_optimizer. I don't think this is desirable, at least not until we validated the correctness and performance equivalence. So, @jomayeri, can you please restore the old lines 1211--1213 and place it above the new lines 1211--1213. In other words, we need both logics. We can chat more for clarifications. Thanks!

Reply:
I thought this was the case, thank you for elaborating, Tunji. 

As I was suggesting a year ago we needed some specific flag in the config that activates `BF16Optimizer`. It keeps on biting us the way it is now.

Schedule a call for Monday? Please ping me on slack when it's good for you